### PR TITLE
Parametrize evaluate

### PR DESCRIPTION
**Pull request by FrancescMartiEscofetQC**
This PR parametrizes the `evaluate` method of the `MetaLearner` so the user can choose which metrics to use to evaluate each individual model. It also implements the method for the `X` and `DR` learners.
# Checklist

- [x] Added a `CHANGELOG.rst` entry

### CODE REVIEW

File: metalearners/metalearner.py [hunk 1]
```diff
@@ -133,6 +135,41 @@ def _validate_n_folds_synchronize(n_folds: dict[str, int]) -> None:
...
+    is_treatment: bool,
+    oos_method: OosMethod = OVERALL,
+) -> dict[str, float]:
+    """Helper function to evaluate all the models of the same model kind."""
+    prefix = f"{model_kind}_"
+    evaluation_metrics: dict[str, float] = {}
+    for idx, scorer in enumerate(scorers):
+        if isinstance(scorer, str):
+            scorer_str = scorer
+            scorer_call: Callable = get_scorer(scorer)
```
Comment (kklein) [inline metalearners/metalearner.py:154]:
```suggestion
            scorer_callable: Callable = get_scorer(scorer)
```

File: metalearners/metalearner.py [hunk 2]
```diff
@@ -133,6 +135,41 @@ def _validate_n_folds_synchronize(n_folds: dict[str, int]) -> None:
...
+    is_oos: bool,
+    is_treatment: bool,
+    oos_method: OosMethod = OVERALL,
+) -> dict[str, float]:
+    """Helper function to evaluate all the models of the same model kind."""
+    prefix = f"{model_kind}_"
+    evaluation_metrics: dict[str, float] = {}
+    for idx, scorer in enumerate(scorers):
+        if isinstance(scorer, str):
+            scorer_str = scorer
```
Comment (kklein) [inline metalearners/metalearner.py:153]:
```suggestion
            scorer_name = scorer
```

File: metalearners/metalearner.py [hunk 3]
```diff
@@ -133,6 +135,41 @@ def _validate_n_folds_synchronize(n_folds: dict[str, int]) -> None:
...
+        for i, cfe in enumerate(cfes):
+            if is_treatment:
+                treatment_variant = i + 1
+                index_str = f"{treatment_variant}_vs_0_"
+            else:
+                if len(cfes) == 1:
+                    index_str = ""
+                else:
+                    index_str = f"{i}_"
+            name = f"{prefix}{index_str}{scorer_str}"
```
Comment (kklein) [inline metalearners/metalearner.py:167]: Did you consciously decide against nesting the dictionary and for concatenating strings?

File: metalearners/metalearner.py [hunk 4]
```diff
@@ -133,6 +135,41 @@ def _validate_n_folds_synchronize(n_folds: dict[str, int]) -> None:
...
+    evaluation_metrics: dict[str, float] = {}
+    for idx, scorer in enumerate(scorers):
+        if isinstance(scorer, str):
+            scorer_str = scorer
+            scorer_call: Callable = get_scorer(scorer)
+        else:
+            scorer_str = f"custom_scorer_{idx}"
+            scorer_call = scorer
+        for i, cfe in enumerate(cfes):
+            if is_treatment:
```
Comment (kklein) [inline metalearners/metalearner.py:159]: I'm not sure I can follow what we're doing here... :/

File: metalearners/metalearner.py [hunk 5]
```diff
@@ -311,6 +348,16 @@ def _validate_models(self) -> None:
...
             )
 
+    @classmethod
+    def _validate_scoring(cls, scoring: Mapping[str, list[str | Callable]]):
+        if not set(scoring.keys()) <= (
+            set(cls.nuisance_model_specifications().keys())
+            | set(cls.treatment_model_specifications().keys())
+        ):
+            raise ValueError(
+                "scoring dict keys need to be a subset of the model names in the MetaLearner"
```
Comment (kklein) [inline metalearners/metalearner.py:358]:
```suggestion
                "keys of parameter scoring need to be a subset of MetaLearner's base model names"
```

Comment (kklein) [reply metalearners/metalearner.py:358]: Why do we need them to be a subset? Why don't we ignore superficial ones?

E.g. if someone is iterating over various MetaLearners and is trying to reuse a scoring dict for several MetaLearners, this won't be possible even though we might just allow for it.

E.g.

```python
prefixes = ["T", "DR"]
for prefix in prefixes:
    metalearner = metalearner_factory(prefix)(
        nuisance_model_factory=LGBMRegressor,
        propensity_model_factory=LGBMClassifier,
        is_classification=False,
        n_variants=2,  
    )
    print(metalearners.evaluate(
        X=X,
        y=y,
        w=w,
        is_oos=False,
        scoring={"propensity_model": "nll", "variant_outcome_model": "rmse"},
    ))
```

File: metalearners/metalearner.py [hunk 6]
```diff
@@ -824,8 +871,39 @@ def evaluate(
         w: Vector,
         is_oos: bool,
         oos_method: OosMethod = OVERALL,
-    ) -> dict[str, float | int]:
-        """Evaluate all models contained in a MetaLearner."""
+        scoring: Mapping[str, list[str | Callable]] | None = None,
+    ) -> dict[str, float]:
+        r"""Evaluate the models models contained in the MetaLearner.
```
Comment (kklein) [inline metalearners/metalearner.py:876]:
```suggestion
        r"""Evaluate the the MetaLearner.
```

**Review by kklein — COMMENTED** [review]
Mostly naming :)

File: metalearners/metalearner.py (hunk 3 above)
Comment (FrancescMartiEscofetQC) [reply metalearners/metalearner.py]: I preferred to return a plain dictionary as then it would be easier to convert it to a `dataframe` if needed (e.g. in #9 ). I would be open to changing it but I also think it would carry unnecessary nesting, for example in the case where we only have one model type, we would have multiple levels but the first one would just have one key.

File: metalearners/metalearner.py (hunk 4 above)
Comment (FrancescMartiEscofetQC) [reply metalearners/metalearner.py:159]: The idea of this `if` is to name the treatment models (which predict the CATE directly) with `"{treatment_variant}_vs_0"` to clearly indicate that it is predicting the CATE. We have to add 1 to `i` as the indexing starts at 0 but the first model is estimating the CATE of variant 1 against 0.

File: metalearners/metalearner.py (hunk 5 above)
Comment (FrancescMartiEscofetQC) [reply metalearners/metalearner.py]: [0150106](https://github.com/Quantco/metalearners/pull/8/commits/01501067b288760b5db9b2857699671541a54a61)

File: metalearners/metalearner.py (hunk 3 above)
Comment (kklein) [reply metalearners/metalearner.py]: I see, sgtm!

File: metalearners/metalearner.py [hunk 7]
```diff
@@ -825,8 +864,40 @@ def evaluate(
...
+        r"""Evaluate the MetaLearner.
+
+        The keys in ``scoring`` which are not a name of a model contained in the MetaLearner
+        will be ignored, for information about this names check
+        :meth:`~metalearners.metalearner.MetaLearner.nuisance_model_specifications` and
+        :meth:`~metalearners.metalearner.MetaLearner.treatment_model_specifications`.
+        The values must be a list of:
+
+        * ``string`` representing a ``sklearn`` scoring method. Check
+          `here <https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values>`__
```
Comment (kklein) [inline metalearners/metalearner.py:878]:
```suggestion
          `here <https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values>`_
```

File: metalearners/metalearner.py [hunk 8]
```diff
@@ -825,8 +864,40 @@ def evaluate(
...
+          for the possible values.
+        * ``Callable`` with signature ``scorer(estimator, X, y_true, **kwargs)``. We recommend
+          using `sklearn.metrics.make_scorer <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html>`_
+          to create this callables.
```
Comment (kklein) [inline metalearners/metalearner.py:882]:
```suggestion
          to create such a ``Callable``.
```

File: metalearners/metalearner.py [hunk 9]
```diff
@@ -133,6 +137,41 @@ def _validate_n_folds_synchronize(n_folds: dict[str, int]) -> None:
...
 
 
+def _evaluate_model_kind(
+    cfes: Sequence[CrossFitEstimator],
+    Xs: Sequence[Matrix],
+    ys: Sequence[Vector],
+    scorers: Sequence[str | Callable],
+    model_kind: str,
+    is_oos: bool,
+    is_treatment: bool,
```
Comment (kklein) [inline metalearners/metalearner.py:147]:
```suggestion
    is_treatment_model: bool,
```
I first thought this had to do with `w=1` or `w=0`. :)

File: metalearners/tlearner.py [hunk 10]
```diff
@@ -114,21 +114,21 @@ def evaluate(
...
-        # TODO: Parametrize evaluation approaches.
-        conditional_average_outcomes = self.predict_conditional_average_outcomes(
-            X=X, is_oos=is_oos, oos_method=oos_method
+        scoring: Scoring | None = None,
+    ) -> dict[str, float]:
+        safe_scoring = self._scoring(scoring)
+
+        masks = []
+        for tv in range(self.n_variants):
+            masks.append(w == tv)
```
Comment (kklein) [inline metalearners/tlearner.py:123]:
```suggestion
```

File: metalearners/xlearner.py [hunk 11]
```diff
@@ -285,9 +287,72 @@ def evaluate(
...
-        raise NotImplementedError(
-            "This feature is not yet implemented for the X-Learner."
+        scoring: Scoring | None = None,
+    ) -> dict[str, float]:
+        safe_scoring = self._scoring(scoring)
+
+        masks = []
+        for tv in range(self.n_variants):
+            masks.append(w == tv)
+
```
Comment (kklein) [inline metalearners/xlearner.py:297]:
```suggestion

```

File: metalearners/metalearner.py (hunk 7 above)
Comment (FrancescMartiEscofetQC) [reply metalearners/metalearner.py:878]: This is to avoid a warning when building the docs that the reference already exists (as there's another link which is reference with `here` also), see [here](https://stackoverflow.com/questions/27420317/restructured-text-rst-http-links-underscore-vs-use).

File: metalearners/drlearner.py [hunk 12]
```diff
@@ -205,37 +208,85 @@ def evaluate(
...
-    ) -> dict[str, float | int]:
-        raise NotImplementedError(
-            "This feature is not yet implemented for the DR-Learner."
+        scoring: Scoring | None = None,
+    ) -> dict[str, float]:
+        safe_scoring = self._scoring(scoring)
+
+        masks = []
+        for tv in range(self.n_variants):
+            masks.append(w == tv)
```
Comment (kklein) [inline metalearners/drlearner.py:217]:
```suggestion
```

File: metalearners/_utils.py [hunk 13]
```diff
@@ -463,3 +463,26 @@ def simplify_output_2d(tensor: np.ndarray) -> np.ndarray:
             "This function requires a regression or a classification with binary outcome "
             "task."
         )
+
+
+# Taken from https://stackoverflow.com/questions/13741998/is-there-a-way-to-let-classes-inherit-the-documentation-of-their-superclass-with
+def copydoc(fromfunc, sep="\n"):
+    """
+    Decorator: Copy the docstring of `fromfunc`
```
Comment (kklein) [inline metalearners/_utils.py:471]:
```suggestion
    Decorator: Copy the docstring of ``fromfunc``
```

**Review by kklein — COMMENTED** [review]
The last `mask`s; I promise!

File: metalearners/drlearner.py (hunk 12 above)
Comment (FrancescMartiEscofetQC) [reply metalearners/drlearner.py:217]: Hahaha, this is my bad, it comes from an old implementation of `evaluate_model` and forgot about them

File: metalearners/rlearner.py [hunk 14]
```diff
@@ -323,22 +326,51 @@ def predict(
...
         y: Vector,
         w: Vector,
         is_oos: bool,
         oos_method: OosMethod = OVERALL,
-    ) -> dict[str, float | int]:
+        scoring: Scoring | None = None,
+    ) -> dict[str, float]:
+        """In the RLearner case, the ``"treatment_model"`` is always evaluated with the
+        :func:`~metalearners.rlearner.r_loss` and the ``scoring["treatment_model"]``
+        parameter is ignored."""
```
Comment (MatthiasLoefflerQC) [inline metalearners/rlearner.py:341]: What if we want to evaluate the treatment model with a different scoring function than the R-loss ? E.g. if the metalearner type is a hyperparameter itself, we might want to use the same evaluation metric across all metalearners and the R-loss favours the RLearner obviously.

Comment (FrancescMartiEscofetQC) [reply metalearners/rlearner.py]: You are right that this should be possible, I adapted the code and now it's possible to evaluate the treatment model with whichever metric the user wants (which supports sample weights, as the model is trained with sample weights) besides the R-Loss.
[c20ae75](https://github.com/Quantco/metalearners/pull/8/commits/c20ae75321c972763c177bf5051f276336eb8924)

**Review by MatthiasLoefflerQC — APPROVED** [review]

**Review by kklein — APPROVED** [review]
